src/envoy/http/authn/http_filter.cc

```diff
@@ -48,54 +49,44 @@ FilterHeadersStatus AuthenticationFilter::decodeHeaders(HeaderMap& headers,
   state_ = State::PROCESSING;
 
   filter_context_.reset(new Istio::AuthN::FilterContext(
-      &headers, decoder_callbacks_->connection(), filter_config_));
+      decoder_callbacks_->requestInfo().dynamicMetadata(),
+      decoder_callbacks_->connection(), filter_config_));
 
   Payload payload;
 
   if (!filter_config_.policy().peer_is_optional() &&
       !createPeerAuthenticator(filter_context_.get())->run(&payload)) {
     rejectRequest("Peer authentication failed.");
-    removeJwtPayloadFromHeaders();
     return FilterHeadersStatus::StopIteration;
   }
 
   bool success =
       filter_config_.policy().origin_is_optional() ||
       createOriginAuthenticator(filter_context_.get())->run(&payload);
 
-  // After Istio authn, the JWT headers consumed by Istio authn should be
-  // removed.
-  removeJwtPayloadFromHeaders();
-
   if (!success) {
     rejectRequest("Origin authentication failed.");
     return FilterHeadersStatus::StopIteration;
   }
 
   // Put authentication result to headers.
   if (filter_context_ != nullptr) {
-    // TODO(yangminzhu): Remove the header and only use the metadata to pass the
+    // TODO(diemvu): Remove the header and only use the metadata to pass the
     // attributes.
     Utils::Authentication::SaveResultToHeader(
-        filter_context_->authenticationResult(), filter_context_->headers());
-
+        filter_context_->authenticationResult(), &headers);
```

Comment:
so we are still using headers to pass data between authn and mixer?

Reply:
Yes. As mentioned in PR description, I want to do it in different PRs to make the change size reasonable.